### DISCUSSION TITLE
Code style: where to put binary operator in multiline java expression in the code

### DISCUSSION THREAD

**Discussion by kokosing** [trinodb/trino#8157] [General]
Where to put binary operator in multiline java expression in the code.

1. 
```
expr1
    && expr2
```

```
$ git grep '^ *&& ' | wc -l
217
```

2.

```
expr1 &&
     expr2
```

```
$ git grep '&&$' | wc -l
1467
```     

I prefer option 1 as it a second line starts with the operator so it is clear that it is continuation. It is similar to:

```
method()
     .with()
     .method()
     .chain()
```

**Comment by hashhar** [comment]
+1 for the first personally. One of the benefits is it's easier to see chains of conditions.

Compare (notice that one of them is different is more easily seen in the first form):

```java
return limitNode.getCount() == limit
        && limitNode.isWithTies() == !tiesResolvers.isEmpty()
        || limitNode.isPartial() == partial
        && limitNode.requiresPreSortedInputs() == !preSortedInputs.isEmpty();
```

```java
return limitNode.getCount() == limit && 
           limitNode.isWithTies() == !tiesResolvers.isEmpty() ||
           limitNode.isPartial() == partial &&
           limitNode.requiresPreSortedInputs() == !preSortedInputs.isEmpty();
```

But it seems like we are mostly using the 2nd form today (by a large margin).

```sh
# Install ripgrep and run
# 1st form
$ rg '^\s+(&&|\|\|)' --type java | wc -l
     329

# 2nd form
$ rg '(&&|\|\|)$' --type java | wc -l
    1613
```

**Comment by kokosing** [comment]
> But it seems like we are mostly using the 2nd form today (by a large margin).

Yeah. But if set a new rule then we can migrate over time to a direction we want.

> Compare (notice that one of them is different):

I think the example code requires a cleanup. Mixing `&&` and `||` together makes it really hard to read.

**Comment by hashhar** [comment]
> I think the example code requires a cleanup. Mixing && and || together makes it really hard to read.

That was the point - with the trailing form it's difficult to notice that something might be wrong. (the original code uses all `&&` btw).

**Comment by findepi** [comment]
What does intellij write when generating equals?

**Comment by kokosing** [comment]
> What does intellij write when generating equals?

For me it does not format this.

```
        return external == that.external && rowId.equals(that.rowId) && scanAuthorizations.equals(that.scanAuthorizations) && schema.equals(that.schema) && serializerClassName.equals(that.serializerClassName) && table.equals(that.table) && constraint.equals(that.constraint);
```

**Comment by findepi** [comment]
Indeed. intellij doesn't output linebreaks for me either. 
A different, but related, question -- how does intellij format the code when i insert a linebreak (by pressing enter) within `"some string literal"`?

**Reply by kokosing** [reply]
For me it generates the below, but I am afraid it could due the way it is configured for me.
```
                        String s = "al" 
                                + "a ma kota";
```

**Reply by findepi** [reply]
for me it generates like this
```
this.tableName = requireNonNull(tableName, "tableName i" +
                "s null");
```
